.circleci/parse_yaml.py

```diff
@@ -0,0 +1,269 @@
+                         ncols=42,
+                         unit_scale=True,
+                         unit_divisor=1024):
+            wrote = wrote + len(data)
+            f.write(data)
+    
+    if total_size != 0 and wrote != total_size:
+        logging.error(f' check the file link and try again.')
+        sys.exit(1)
+
+
+def validate_hash(path, hash):
+    """Validate a file against an MD5 hash value."""
+    import hashlib
+
+    md5 = hashlib.md5()
+    
+    with open(path, 'rb') as f:
+        while True:
+            chunk = f.read(10000000)
+            if not chunk:
+                break
+            md5.update(chunk)
+    
+    return md5.hexdigest() == hash
+
+
+class Yamler:
+    """A collection of utilities to parse the yaml file, check metadata
+    and trigger cimr processing of the contributed file
+    """
+
+    def __init__(self, yaml_data):
+        self.yaml_data = yaml_data
+        self.data_type = None
+        self.keys = None
+        self.hash = None
+        self.outdir = None
+        self.downloaded_file = None
+
+
+    def pick_keys(self):
+        """List keys for the dictionarized yaml data and store in self.
+        The following keys are expected:
+        ['defined_as', 'data_file', 'contributor', 'data_info', 'method']
+        """
+        self.keys = self.yaml_data.keys()
+
+
+    def set_data_type(self):
+        """One of the following data_type variables are expected:
+        ['gwas', 'twas', 'eqtl', 'sqtl', 'pqtl', 'tad']
+        """
+        if self.yaml_data['data_info']['data_type'] is not None:
+            self.data_type = self.yaml_data['data_info']['data_type']
+        else:
+            logging.error(f' there is no data_type indicated.')
+            sys.exit(1)
+
+
+    def download(self):
+        """Check if provided weblink to the file exists. 
+        Download if verified.
+        """
+        path = self.yaml_data['data_file']['location']['url']
+        self.downloaded_file = path.split('/')[-1]
+
+        outdir_root = 'submitted_data/'
+        pathlib.Path(outdir_root).mkdir(exist_ok=True)
+        self.outdir = outdir_root + str(self.data_type) + '/'
+        pathlib.Path(self.outdir).mkdir(exist_ok=True)
+
+        if verify_weblink(path):
+            logging.info(f' starting download')
+            download_file(path, self.outdir)
+            self.hash = self.yaml_data['data_file']['location']['md5']
+        else:
+            logging.error(f' file unavailable')
+            sys.exit(1)
+
+
+    def bulk_download(self):
+        """Bulk download option assumes one of the following file types:
+        ['tgz', 'tar.gz', 'tar.bz2', 'tar.xz']
+        """
+        import tarfile
+
+        self.download()
+
+        if tarfile.is_tarfile(self.downloaded_file):
+            tarred_data = tarfile.open(
+                self.downloaded_file, 
+                mode='r:*'
+            )
+            tarred_data.extractall(path=self.outdir)
```

Comment:
One question on `extractall` method: When the tarball file is extracted by this method, does it put the archived content in the directory where the tarball file is located?

Reply:
After the tarball file is extracted, should we also check the archived file system structure to ensure that it always follows the format of `<data_type>/<filename>`? That will simplify `.circleci/process_submitted_data.py`.

Reply:
re: extractall - `path` argument defines the directory where the contents of the tarball will be placed. 
re: outdir - Currently, the way I am describing it in the documents is that the tarfile should not contain any directory trees. However, yes, it would make sense to include some checks here. I will add something.